# Profitable App Profiles for the App Store and Google Play Markets

As an app developer company, we make our apps available on google apps and App store. Our main source of revenue is in - app adds. 

Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.


<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Profitable-App-Profiles-for-the-App-Store-and-Google-Play-Markets" data-toc-modified-id="Profitable-App-Profiles-for-the-App-Store-and-Google-Play-Markets-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Profitable App Profiles for the App Store and Google Play Markets</a></span><ul class="toc-item"><li><span><a href="#Data-source" data-toc-modified-id="Data-source-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Data source</a></span></li><li><span><a href="#Opening-and-exploring-the-data" data-toc-modified-id="Opening-and-exploring-the-data-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Opening and exploring the data</a></span></li><li><span><a href="#Deleting-wrong-data" data-toc-modified-id="Deleting-wrong-data-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Deleting wrong data</a></span></li><li><span><a href="#Deleting-duplicate-entries" data-toc-modified-id="Deleting-duplicate-entries-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Deleting duplicate entries</a></span></li><li><span><a href="#Removing-non-english-apps-(Part-1)" data-toc-modified-id="Removing-non-english-apps-(Part-1)-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Removing non-english apps (Part 1)</a></span></li><li><span><a href="#Isolating-Free-Apps" data-toc-modified-id="Isolating-Free-Apps-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Isolating Free Apps</a></span></li><li><span><a href="#Most-common-apps-by-genre" data-toc-modified-id="Most-common-apps-by-genre-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Most common apps by genre</a></span></li><li><span><a href="#Most-common-apps-by-genre-:-Part-2" data-toc-modified-id="Most-common-apps-by-genre-:-Part-2-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Most common apps by genre : Part 2</a></span></li><li><span><a href="#Most-Common-Apps-By-genre-:-Part-3" data-toc-modified-id="Most-Common-Apps-By-genre-:-Part-3-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Most Common Apps By genre : Part 3</a></span></li><li><span><a href="#Most-popular-apps-by-genre-on-the-AppStore-(IOS-Apps)" data-toc-modified-id="Most-popular-apps-by-genre-on-the-AppStore-(IOS-Apps)-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Most popular apps by genre on the AppStore (IOS Apps)</a></span></li><li><span><a href="#Most-popular-apps-by-genre-on-Google-Play-(Android-Apps)" data-toc-modified-id="Most-popular-apps-by-genre-on-Google-Play-(Android-Apps)-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>Most popular apps by genre on Google Play (Android Apps)</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li></ul></div>

## Data source 

We have collected the data from two source :
* Google play store data [at this site](https://www.kaggle.com/lava18/google-play-store-apps)
* App Store data [at this site](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

## Opening and exploring the data

In [63]:
from csv import reader
# opening Android data
opened_file = open('googleplaystore.csv', encoding="utf8")
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

### The App Store data set ###
opened_file = open('AppleStore.csv', encoding = "utf8")
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]


To make it easier to explore the two data sets, we'll first write a function named explore_data() that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set

In [64]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


Lets explore the two datasets. 

In [65]:
print(android_header) # name of all columns of android data sets 


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [66]:
explore_data(android,0,3, True) # showing first 3 rows

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [67]:
print(ios_header) # name of all columns of ios data sets 

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [68]:
explore_data(ios,0,3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


**We have 10841 observations for android apps and 7197 observations for IOS apps.** For android apps , the columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

For IOS apps, 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre' are the columns we would focus on.

## Deleting wrong data

Before beginning our analysis, we need to make sure the data we analyze is accurate, otherwise the results of our analysis will be wrong.

we need to:

* Detect inaccurate data, and correct or remove it.
* Detect duplicate data, and remove the duplicates.

We only build apps that are free to download and install, and that are directed toward an English-speaking audience. This means that we'll need to:

* Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.
* Remove apps that aren't free.

In [one discussion in the data source page](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015), it says that the column 10472 has a wrong value in the rating section. Lets find it out.

In [69]:
print(android[10472]) # the wrong data column
print(android_header) # the headers

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


It says that the rating is 19 ! but the rating can not be more than 5. So we will remove this row. 

In [70]:
print (len(android)) # number of rows before deletion

10841


In [71]:
del android[10472]

In [72]:
print(len(android)) # checking the number of rows after the deletion

10840


## Deleting duplicate entries

We will remove all the apps that have more than one entries. This very interesting piece of code helps us easily detect the duplicate entires. Here we make two blank lists. Unique and duplicate. For each iteration in the app name column, the code checks and sorts the duplicate rows and unique rows.

In [73]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else :
        unique_apps.append(name)
        
print(len(duplicate_apps))


1181


Now the nest step is remove the duplicates. 

If you examine the rows we printed two cells above for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

To do that, we will:

* Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
* Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)



In [74]:
reviews_max = {} # empty dictionary

for app in android:
    name = app[0]
    n_reviews = float(app[3]) # number of reviews variable stores the review column in the data.
    
    if name in reviews_max and reviews_max[name] < n_reviews: # if the review value for that particular entry is less than n_reviews 
        reviews_max[name] = n_reviews # update the value of that entry according to the n_review value of that entry(making them equal), by this, ensuring the max number of the review being recorded.
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews # If name is not in the reviews_max dictionary as a key, create a new entry in the dictionary where the key is the app name, and the value is the number of reviews

In [75]:
print(len(reviews_max))
print(reviews_max[name])# if you are wondering what is 398307, its not the length of the dictionary, its the value of the last item in the dictionary. Check exampe code below.

9659
398307.0


In [76]:
# EXMAPLE FOR UNDERSTANDING

check = {'a':1,'b':2,'c':20}
for n in check:
    print(n)
print(len(check))
print(check[n]) # the last entry value
len(check) == check[n] # of course they are not equal

a
b
c
3
20


False

So, in the above code cell, we seperated the entires with maximum reviews in a list called reviews_max. We made sure that if any entry belongs in the android data set, **only** the highest review of that entry is saved in a seperated dictionary. **Thus making a set of unique values.**

Now, as we have only the unique apps stored with the highest reviews in **reviews_max** dict, the difference between the android set and duplicate set should be equal to the number of entries in reviews_max dictionary. Lets find out.

In [77]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

1. We start by initializing two empty lists, android_clean and already_added.
2. We loop through the android data set, and for every iteration:
3. We isolate the name of the app and the number of reviews.
4. We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
    * The number of reviews of the current app matches the number of reviews of       that app as described in the reviews_max dictionary; and
    * The name of the app is not already in the already_added list. 
Already added list can contain a entry multiple times if a particular entry has the same number of reviews recorded in the android set couple of times. We need to add this second supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [78]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added): #if the value of the particular entry in the review_max dict = n_reviews
        android_clean.append(app)
        already_added.append(name) # make sure this is inside the if block

Lets check the android_clean dataset, with our explore_data function

In [79]:
explore_data(android_clean,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


yes, the number of rows is 9659. Which is equal to the actual set minus the duplicated entries.

## Removing non-english apps (Part 1)

We use English for the apps we develop at our company, and we'd like to analyze only the apps that are directed toward an English-speaking audience. However, if we explore the data long enough, we'll find that both data sets have apps with names that suggest they are not directed toward an English-speaking audience.

We can detect any other language in app names using ASCII codes. English represents the number range from 0 to 127. We can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

We will build a function that can detect whether a particular string belongs to English or not. The **ord** function can detect the particular code of a string. It if detects the code is more than 127, it would should return false in the English string detection function.

In [80]:
def is_english(string):
    
    for character in string:
        if ord(character) > 127: #if ord of the character is greater than 127
            return False
    
    else :
        return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

In [81]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

print(ord('™'))
print(ord('😜'))

False
False
8482
128540


We can see that, even if the first two apps are English, our function is detecting it as non-english app. We can modify our function to allow at least 3 characters whose **ord** values are more than 127. This will somewhat reduce the possibility to leave out English apps with symbols and other emojis. 

In [82]:
def is_english(string):
    non_ascii = 0 # declaring a variable to store characters which ord value are more that 127
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

# Lets check 

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
True


Now, at least its not detecting those apps as non - english. Now , we will create two new lists or datasets which consists of English android and IOS apps. 

In [83]:
android_english = [] 
ios_english = []

for app in android_clean : # for every row in android_clean dataset
    name = app[0] # the name column is the first row
    if is_english(name) : # if our function detects the name as english
        android_english.append(app) # store the whole row of that particular app in android_english dataset
    
# Same for IOS apps     
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
                

Now lets explore those two datasets.

In [84]:
explore_data(android_english,0,3,True)
print('\n') # just creating a gap between two outputs
explore_data(ios_english,0,3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

**Total English apps for Android** : 9614

**Total English apps for IOS** : 6183

## Isolating Free Apps

We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis. 



In [85]:
# Looking at the columns again to check which one shows the price 

print(android_header)
print('\n')
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


The 7th column shows the price for android apps and the 4th column shows the price for IOS apps.

In [86]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print('The number of free android apps :', len(android_final))
print('The number of free IOS apps :',len(ios_final))

The number of free android apps : 8864
The number of free IOS apps : 3222


## Most common apps by genre

Our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

* Build a minimal Android version of the app, and add it to Google Play.
* If the app has a good response from users, we develop it further.
* If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. 

**Let's begin the analysis by getting a sense of what are the most common genres for each market.** For this, we'll need to build frequency tables for a few columns in our data sets.

For this, we'll build a frequency table for the `prime_genre` column of the App Store data set, and the `Genres` and `Category` columns of the Google Play data set

## Most common apps by genre : Part 2

We'll build two functions we can use to analyze the frequency tables:

* One function to generate frequency tables that show percentages
* Another function we can use to display the percentages in a descending order.

This process is a bit complicated to understand. When we are talking about creating a frequency table, we are talking about creating a dictionary. 

Now, dictionaries are difficult to get sorted with the built in **soreted()** function. We have to convert the dictionary to tuples and then sort the tuple in descending order to generate a frequency table. 

First, We will make the function that shows the percentages. 

In [87]:

def freq_table(dataset, index): # 
    table = {} # creating empty dict which will store the genre or categories
    total = 0
    
    for row in dataset: # for every row
        total += 1 # count the row
        value = row[index] # value stores the genre or category name
        if value in table: # if that genre is already stored in the table dict
            table[value] += 1 # store one more instance of it
        else:
            table[value] = 1 # else, create a new instance of the genre
    
    table_percentages = {} # empty dictionary
    for key in table: # for every genre name as a key in the table
        percentage = (table[key] / total) * 100 # number of that genre divided by the total count of the row
        table_percentages[key] = percentage # table_percentage dictionary will be filled by that genre or category and their respective percentage
    
    return table_percentages # return the accumulated dictionary / freq percentage


def display_table(dataset, index):
    table = freq_table(dataset, index) # storing all the genre and their percentages(as a dict) in the table variable
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key) # for every key in the dictionary, store the value of that dict in the first position and the genre name as the second position in a tuple
        table_display.append(key_val_as_tuple) # store this tuple in the table_display list
        
    table_sorted = sorted(table_display, reverse = True) # sorting the freq table in descending order
    for entry in table_sorted:
        print(entry[1], ':', entry[0]) # for every genre in the table, print the genre name first and then the percentage

## Most Common Apps By genre : Part 3 

We will use our function to see what does it tell us about the free **English** apps in AppStore. 

In [88]:
display_table(ios_final,-5) # -5 is the index number of prime_genre column, 5th position from the last

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We see the most number of apps in the Appstore are game apps, followed by entertainmainment, social networking and education apps. However, this doesnt mean that games apps have most number of users. It just says that the highest number of apps in Appstore are game apps

In [89]:
display_table(android_final, -4) # most number of android apps sorted by genre

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In [90]:
display_table(android_final, 1) # most number of android apps sorted by category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

At google play, most number of apps belongs to the family category. But after visitng the google play store's family category, we have seen that mostly games for kids are in family category. 

If we see the frequency tables by genre and category , we see that most number of games are tools, game and entertainment. 

Although if we see, sorting by **genres**, gives us names of more types of apps than we got while sorting by categories. However, as we will only work with the tops and not going in very detail, we will work with the **category** column from android_final dataset. 



## Most popular apps by genre on the AppStore (IOS Apps)

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [91]:
genres_ios = freq_table(ios_final, -5) # All the genres in a table
for genre in genres_ios:
    total = 0 
    len_genre = 0
    for app in ios_final: # for every app in the ios_final list
        genre_app = app[-5] # put the genre name(app[5]) in genre_app variable
        if genre_app == genre:  # if the name of the app matches with genre var          
            n_ratings = float(app[5]) # store the rating (rating_count_tot) in n_ratings
            total += n_ratings # increament the total var with n_ratings
            len_genre += 1 # increment the len_genre by 1
    avg_n_ratings = total / len_genre # store the avg rating by summing up the total ratings divided by the total number of ratings
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


On average, navigation apps have the highest number of user reviews. But is it really the case that 'Navigation' apps have the highest number of users ?

Lets check all apps in the Navigation genre.

In [92]:

for app in ios_final:
    if app[-5] == 'Navigation': # if app's category column says its 'Navigation'
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


Now this is important. While we are working to find the most popular genre of apps in AppStore by averaging total number of ratings, the average number of ratings might be influenced by only the outliers, like Google maps. This will not allow us to see the actual average picture that we want to see. 

Competing against apps like google maps wont be an effective decision because most of the market is heavily relying towards it. We have to compete with the genres that where there is not a particular app that is ruling the market. There we will have a better chance of having a 'level playing field' to compete with. 

Another way out is to investigate more popular app genres and find out which genre have potential to add features with less effort. Lets investigate some popular apps. Music and reference seems like the genres we want to investigate in this case. 

Music : 57326.530303030304

Reference : 74942.11111111111


In [93]:
for app in ios_final:
    if app[-5] == 'Music': # if app's category column says its 'Navigation'
        print(app[1], ':', app[5])

Pandora - Music & Radio : 1126879
Spotify Music : 878563
Shazam - Discover music, artists, videos & lyrics : 402925
iHeartRadio – Free Music & Radio Stations : 293228
SoundCloud - Music & Audio : 135744
Magic Piano by Smule : 131695
Smule Sing! : 119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music : 110420
Amazon Music : 106235
SoundHound Song Search & Music Player : 82602
Sonos Controller : 48905
Bandsintown Concerts : 30845
Karaoke - Sing Karaoke, Unlimited Songs! : 28606
My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398
YouTube Mu

Music apps can be investigated further if there is any additional features or demand that the market is missing. But one thing is for sure ; we have to put a lot of errort in the production, because there are some technical things like artist permission and other legal stuffs involved. Also involved the royality. 

So, music might not be a good case for us. Lets investigate the **Reference** genre.

In [94]:
for app in ios_final:
    if app[-5] == 'Reference': # if app's category column says its 'Navigation'
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Here, most popular books and references are Bible and dictionaries. This seems like a good option for us to delve in. Because, most popular books tend to be in public domain. So if we make app versions of those books, we might not have to think about issues like royality and legal. 

We can add some interesting features by understanding user demands by researching opinions and feedbacks.  

***Thinking more granular questions = thinking more possibility***

## Most popular apps by genre on Google Play (Android Apps) 

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [95]:
display_table(android_final, 5) # 5 is the 'installs' column.

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error.

In [96]:
categories_android = freq_table(android_final, 1) # storing all the categries

print(categories_android)

{'ART_AND_DESIGN': 0.6430505415162455, 'AUTO_AND_VEHICLES': 0.9250902527075812, 'BEAUTY': 0.5979241877256317, 'BOOKS_AND_REFERENCE': 2.1435018050541514, 'BUSINESS': 4.591606498194946, 'COMICS': 0.6204873646209386, 'COMMUNICATION': 3.2378158844765346, 'DATING': 1.861462093862816, 'EDUCATION': 1.1620036101083033, 'ENTERTAINMENT': 0.9589350180505415, 'EVENTS': 0.7107400722021661, 'FINANCE': 3.7003610108303246, 'FOOD_AND_DRINK': 1.2409747292418771, 'HEALTH_AND_FITNESS': 3.0798736462093865, 'HOUSE_AND_HOME': 0.8235559566787004, 'LIBRARIES_AND_DEMO': 0.9363718411552346, 'LIFESTYLE': 3.9034296028880866, 'GAME': 9.724729241877256, 'FAMILY': 18.907942238267147, 'MEDICAL': 3.531137184115524, 'SOCIAL': 2.6624548736462095, 'SHOPPING': 2.2450361010830324, 'PHOTOGRAPHY': 2.944494584837545, 'SPORTS': 3.395758122743682, 'TRAVEL_AND_LOCAL': 2.33528880866426, 'TOOLS': 8.461191335740072, 'PERSONALIZATION': 3.3167870036101084, 'PRODUCTIVITY': 3.892148014440433, 'PARENTING': 0.6543321299638989, 'WEATHER': 

In [97]:
for category in categories_android: # for each category
    total = 0
    len_category = 0 
    for app in android_final: # for each app
        category_app = app[1]  # store the name of the category extracted from each row in the android_final dataset
        if category_app == category: # if the category name matches with the 'category' from category_android          
            n_installs = app[5] # store number of installs from the 'install' column
            n_installs = n_installs.replace(',', '') # replace the commas with blank
            n_installs = n_installs.replace('+', '') # replace the + sign with blank
            total += float(n_installs) # increment the total var with number of installs
            len_category += 1 # add 1 to the number of categories variable
    avg_n_installs = total / len_category # get the avg
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs. Lets see the communication apps and their number of installs. 

In [98]:
for app in android_final :
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

In the communication category, there are many kinds of apps like messenger, vdo, browser etc. We would want to narrow our focus on some apps and would like to see the average number of installs for popular android apps. 

But if we want to take average from above numbers, we would end up having an average which is influenced by **outliers**, thus we wont have the correct picture. 

We will remove these outliers and see the real average and which apps belong to around that average.

In [109]:

under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)


3603485.3884615386

We can see that this (*true average*)  number is much less than other category like books. So, if we remove the outliers, communication category is not really ruling the market. 

Lets investigate the category which have the second highest install : vdo players

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The **books and reference** genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs: 

In [110]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

We can see that there are outliers who will be affecting the average. We would want to see the apps in the books category which falls below the outliers. 

In [113]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [116]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

We see that there are various kinds of apps in this category. It doesnt seem to build another book or reference app in this **crowd** and get lost. 

However, there are plenty of dictionaries which have 10,000,000 + downloads. Another option is to make app version of popular books like we recommended in IOS apps. Now, we have to think which will take more effort.

In any way, we have to add new and interesting features that other apps dont have. 


## Conclusion 

In this project, we tried to find the best popular categories for both **Android** and **IOS** markets. 

We recommend that we should focus on building popular books as apps for Appstore and make dictionaries or popular books for Play store. In any ways, we suggest that new features have to be added to stand out in the competetion. Features like audio, creating case studies on particular chapters etc can pull more users in the market. 